### ISSUE TITLE
Custom Loss Function feature

### ISSUE THREAD

**Issue by ericleonardo** [Nixtla/neuralforecast#669]
### Description

Please, can you add Custom Loss Function feature?
So anyone can build and use their own loss function.

I asked ChatGPT to make a Number Signal (positive/negative) Mean Accuracy loss function for Pytorch.
Predicting financial/stocks time series is more about the correct Direction (up/down) than point accuracy (RMSE).
Predicting Stocks time series we use returns (t, t-1 difference) to make it stationary and predict if next difference is positive/negative (up/down movement).

If NeuralForecast accept custom loss function, I would use the script suggested by ChatGPT (not tested yet):

    import torch
    import torch.nn as nn

    class NumberSignalLoss(nn.Module):
    def __init__(self):
        super(NumberSignalLoss, self).__init__()

    def forward(self, predictions, targets):
        predictions = torch.sign(predictions)  # Convert predictions to either -1 or 1
        targets = torch.sign(targets)  # Convert targets to either -1 or 1

        accuracy = torch.mean(torch.eq(predictions, targets).float())  # Compute mean accuracy

        loss = 1 - accuracy  # Calculate loss as 1 - accuracy

        return loss

ChatGPT said:
In this example, the NumberSignalLoss class is a subclass of nn.Module that implements the loss function. The forward method takes in two inputs: predictions and targets. Both inputs are assumed to be tensors.

The forward method first applies the torch.sign function to both the predictions and targets. This function converts positive values to 1 and negative values to -1, making the predictions and targets binary signals.

Then, it calculates the accuracy by comparing the element-wise equality of the predictions and targets using torch.eq. The result is then converted to a float tensor and the mean is taken using torch.mean.

Finally, the loss is calculated as 1 - accuracy since we want to minimize the loss. The loss tensor is returned as the output.

You can use this loss function in your PyTorch training loop by instantiating an instance of NumberSignalLoss and passing your predictions and targets to it. The resulting loss can then be used for backpropagation and optimization.

**Comment by cchallu** [comment]
Hi @ericleonardo! Thanks for the suggestion.

Neuralforecast already allows for custom loss functions. All models have a `loss` hyperparameter, which you can use to specify your custom loss.

Note that they will expect it to be a class with specific methods. For reference, see the `MAE` loss  in https://github.com/Nixtla/neuralforecast/blob/main/nbs/losses.pytorch.ipynb.

Here is the (untested) `NumberSignalLoss` in NeuralForecast format. **You will still need to modify the loss to consider the `mask` tensor to avoid using datapoints with missing values**.

```
class NumberSignalLoss(nn.Module):
  def __init__(self):
          super(NumberSignalLoss, self).__init__()
          self.outputsize_multiplier = 1
          self.output_names = ['']
          self.is_distribution_output = False
  
  def domain_map(self, y_hat: torch.Tensor):
          """
          Univariate loss operates in dimension [B,T,H]/[B,H]
          This changes the network's output from [B,H,1]->[B,H]
          """
          return y_hat.squeeze(-1)
  
  def __call__(self, y: torch.Tensor, y_hat: torch.Tensor,
                   mask: Union[torch.Tensor, None] = None):
          """
          **Parameters:**
          `y`: tensor, Actual values.
          `y_hat`: tensor, Predicted values.
          `mask`: tensor, Specifies date stamps per serie to consider in loss.
  
          **Returns:**
          `signal_loss`: tensor (single value).
          """
          y_hat = torch.sign(y_hat)  # Convert predictions to either -1 or 1
          y = torch.sign(targets)  # Convert targets to either -1 or 1
          accuracy = torch.mean(torch.eq(y_hat, y).float())  # Compute mean accuracy
          loss = 1 - accuracy  # Calculate loss as 1 - accuracy
          return loss
```

This loss can be very useful for many other settings. @ericleonardo do you want to try to add it to the library and be an official contributor? If so, I can send you our latest contributing guide.

**Comment by ericleonardo** [comment]
Thank you so much, Cristian! Much appreciated your help.
Please, feel free to to become the official contributor of this loss function if you want.
I'm not experienced in doing it, also you helped a lot adapting it and co-worked on it.
I'm going to test it and give you a feedback. ; )

**Comment by kdgutier** [comment]
We are looking to expand NeuralForecast's contributors. 
We would love it if you could make the `NumberSignalLoss` contribution 🙏, @ericleonardo.

1. `pip install nbdev`
2. Modify the [pytorch losses](https://github.com/Nixtla/neuralforecast/blob/main/nbs/losses.pytorch.ipynb) notebook to include the `NumberSignalLoss`. 
2.1 I suggest adding `NumberSignalLoss` after `MASE` along with the scale independent losses.
2.2 Add documentation with `show_doc(NumberSignalLoss, name='NumberSignalLoss.__init__', title_level=3)`
3. Follow nbdev instructions to make PR:
3.1 Install git hooks `nbdev_install_hooks`
3.2 Translate notebook to .py code `nbdev_export`
3.3 Clean notebooks metadata `nbdev_clean --clear_all` + avoid overwriting tutorials `git checkout nbs/examples`
3.4 PR

Previous detailed instructions to be updated: https://github.com/Nixtla/neuralforecast/blob/main/CONTRIBUTING.md

We will be attentive if you need help.

**Comment by ericleonardo** [comment]
Ok, @kdgutier @cchallu 
I will test the loss function before.
It is my first time with NeuralForecast and I'm very excited with it.
Thank you for the support ; )
